File: src/backend/compras/src/model/compra_requisicao_model.py
from src import db
from src.model.compra_requisicao_detalhe_model import CompraRequisicaoDetalheModel
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel
from src.model.compra_tipo_requisicao_model import CompraTipoRequisicaoModel


class CompraRequisicaoModel(db.Model):
    __tablename__ = 'compra_requisicao'

    id = db.Column(db.Integer, primary_key=True)
    descricao = db.Column(db.String(100))
    data_requisicao = db.Column(db.DateTime)
    observacao = db.Column(db.Text)
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))
    id_compra_tipo_requisicao = db.Column(db.Integer, db.ForeignKey('compra_tipo_requisicao.id'))

    compra_requisicao_detalhe_model_list = db.relationship('CompraRequisicaoDetalheModel', lazy='dynamic')
    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])
    compra_tipo_requisicao_model = db.relationship('CompraTipoRequisicaoModel', foreign_keys=[id_compra_tipo_requisicao])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_compra_tipo_requisicao = data.get('idCompraTipoRequisicao')
        self.id_colaborador = data.get('idColaborador')
        self.descricao = data.get('descricao')
        self.data_requisicao = data.get('dataRequisicao')
        self.observacao = data.get('observacao')

    def serialize(self):
        return {
            'id': self.id,
            'idCompraTipoRequisicao': self.id_compra_tipo_requisicao,
            'idColaborador': self.id_colaborador,
            'descricao': self.descricao,
            'dataRequisicao': self.data_requisicao.isoformat(),
            'observacao': self.observacao,
            'compraRequisicaoDetalheModelList': [compra_requisicao_detalhe_model.serialize() for compra_requisicao_detalhe_model in self.compra_requisicao_detalhe_model_list],
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
            'compraTipoRequisicaoModel': self.compra_tipo_requisicao_model.serialize() if self.compra_tipo_requisicao_model else None,
        }

File: frontend/compras/lib/app/data/model/compra_requisicao_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:compras/app/infra/infra_imports.dart';
import 'package:compras/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class CompraRequisicaoModel {
	int? id;
	int? idCompraTipoRequisicao;
	int? idColaborador;
	String? descricao;
	DateTime? dataRequisicao;
	String? observacao;
	List<CompraRequisicaoDetalheModel>? compraRequisicaoDetalheModelList;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;
	CompraTipoRequisicaoModel? compraTipoRequisicaoModel;

	CompraRequisicaoModel({
		this.id,
		this.idCompraTipoRequisicao,
		this.idColaborador,
		this.descricao,
		this.dataRequisicao,
		this.observacao,
		this.compraRequisicaoDetalheModelList,
		this.viewPessoaColaboradorModel,
		this.compraTipoRequisicaoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'descricao',
		'data_requisicao',
		'observacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Descricao',
		'Data Requisicao',
		'Observacao',
	];

	CompraRequisicaoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCompraTipoRequisicao = jsonData['idCompraTipoRequisicao'];
		idColaborador = jsonData['idColaborador'];
		descricao = jsonData['descricao'];
		dataRequisicao = jsonData['dataRequisicao'] != null ? DateTime.tryParse(jsonData['dataRequisicao']) : null;
		observacao = jsonData['observacao'];
		compraRequisicaoDetalheModelList = (jsonData['compraRequisicaoDetalheModelList'] as Iterable?)?.map((m) => CompraRequisicaoDetalheModel.fromJson(m)).toList() ?? [];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
		compraTipoRequisicaoModel = jsonData['compraTipoRequisicaoModel'] == null ? CompraTipoRequisicaoModel() : CompraTipoRequisicaoModel.fromJson(jsonData['compraTipoRequisicaoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCompraTipoRequisicao'] = idCompraTipoRequisicao != 0 ? idCompraTipoRequisicao : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['descricao'] = descricao;
		jsonData['dataRequisicao'] = dataRequisicao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataRequisicao!) : null;
		jsonData['observacao'] = observacao;
		
		var compraRequisicaoDetalheModelLocalList = []; 
		for (CompraRequisicaoDetalheModel object in compraRequisicaoDetalheModelList ?? []) { 
			compraRequisicaoDetalheModelLocalList.add(object.toJson); 
		}
		jsonData['compraRequisicaoDetalheModelList'] = compraRequisicaoDetalheModelLocalList;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
		jsonData['compraTipoRequisicaoModel'] = compraTipoRequisicaoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCompraTipoRequisicao = plutoRow.cells['idCompraTipoRequisicao']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		descricao = plutoRow.cells['descricao']?.value;
		dataRequisicao = Util.stringToDate(plutoRow.cells['dataRequisicao']?.value);
		observacao = plutoRow.cells['observacao']?.value;
		compraRequisicaoDetalheModelList = [];
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
		compraTipoRequisicaoModel = CompraTipoRequisicaoModel();
		compraTipoRequisicaoModel?.nome = plutoRow.cells['compraTipoRequisicaoModel']?.value;
	}	

	CompraRequisicaoModel clone() {
		return CompraRequisicaoModel(
			id: id,
			idCompraTipoRequisicao: idCompraTipoRequisicao,
			idColaborador: idColaborador,
			descricao: descricao,
			dataRequisicao: dataRequisicao,
			observacao: observacao,
			compraRequisicaoDetalheModelList: compraRequisicaoDetalheModelListClone(compraRequisicaoDetalheModelList!),
		);			
	}

	compraRequisicaoDetalheModelListClone(List<CompraRequisicaoDetalheModel> compraRequisicaoDetalheModelList) { 
		List<CompraRequisicaoDetalheModel> resultList = [];
		for (var compraRequisicaoDetalheModel in compraRequisicaoDetalheModelList) {
			resultList.add(
				CompraRequisicaoDetalheModel(
					id: compraRequisicaoDetalheModel.id,
					idCompraRequisicao: compraRequisicaoDetalheModel.idCompraRequisicao,
					idProduto: compraRequisicaoDetalheModel.idProduto,
					quantidade: compraRequisicaoDetalheModel.quantidade,
				)
			);
		}
		return resultList;
	}

	
}